File: internal/forms/federal/form_8949.go
package federal

import (
	"taxpilot/internal/forms"
)

func init() { forms.RegisterForm(Form8949) }

// Form8949 returns the FormDef for Form 8949 — Sales and Other Dispositions
// of Capital Assets. This form aggregates 1099-B transactions and computes
// gain/loss for each category, feeding into Schedule D.
//
// The IRS splits Form 8949 into two parts:
//   Part I:  Short-term (held ≤ 1 year)
//   Part II: Long-term (held > 1 year)
//
// Each part has three boxes based on whether basis was reported to IRS:
//   Box A/D: Basis reported (no adjustments needed)
//   Box B/E: Basis not reported
//   Box C/F: Cannot determine / Form 1099-B not received
//
// For the MVP, we aggregate all 1099-B transactions into short-term and
// long-term totals. The user indicates term and basis reporting on each 1099-B.
func Form8949() *forms.FormDef {
	return &forms.FormDef{
		ID:           forms.FormF8949,
		Name:         "Form 8949 — Sales and Other Dispositions of Capital Assets",
		Jurisdiction: forms.Federal,
		TaxYears:      []int{2025},
		QuestionGroup: forms.GroupIncome1099,
		QuestionOrder: 3,
		Fields: []forms.FieldDef{
			// --- Part I: Short-Term ---

			// Total short-term proceeds (sum of all short-term 1099-B proceeds)
			{
				Line:      "st_proceeds",
				Type:      forms.Computed,
				Label:     "Short-term total proceeds",
				DependsOn: []string{forms.F1099BWildcardProceeds, forms.F1099BWildcardTerm},
				Compute: func(dv forms.DepValues) float64 {
					return dv.SumAllWhere(forms.F1099BWildcardProceeds, forms.F1099BWildcardTerm, "short")
				},
			},
			// Total short-term cost basis
			{
				Line:      "st_basis",
				Type:      forms.Computed,
				Label:     "Short-term total cost basis",
				DependsOn: []string{forms.F1099BWildcardBasis, forms.F1099BWildcardTerm},
				Compute: func(dv forms.DepValues) float64 {
					return dv.SumAllWhere(forms.F1099BWildcardBasis, forms.F1099BWildcardTerm, "short")
				},
			},
			// Total short-term wash sale adjustments
			{
				Line:      "st_wash",
				Type:      forms.Computed,
				Label:     "Short-term wash sale adjustments",
				DependsOn: []string{forms.F1099BWildcardWashSale, forms.F1099BWildcardTerm},
				Compute: func(dv forms.DepValues) float64 {
					return dv.SumAllWhere(forms.F1099BWildcardWashSale, forms.F1099BWildcardTerm, "short")
				},
			},
			// Short-term gain or loss
			{
				Line:      "st_gain_loss",
				Type:      forms.Computed,
				Label:     "Short-term gain or (loss)",
				DependsOn: []string{forms.F8949STProceedsKey, forms.F8949STBasisKey, forms.F8949STWashKey},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.F8949STProceedsKey) - dv.Get(forms.F8949STBasisKey) + dv.Get(forms.F8949STWashKey)
				},
			},

			// --- Part II: Long-Term ---

			// Total long-term proceeds
			{
				Line:      "lt_proceeds",
				Type:      forms.Computed,
				Label:     "Long-term total proceeds",
				DependsOn: []string{forms.F1099BWildcardProceeds, forms.F1099BWildcardTerm},
				Compute: func(dv forms.DepValues) float64 {
					return dv.SumAllWhere(forms.F1099BWildcardProceeds, forms.F1099BWildcardTerm, "long")
				},
			},
			// Total long-term cost basis
			{
				Line:      "lt_basis",
				Type:      forms.Computed,
				Label:     "Long-term total cost basis",
				DependsOn: []string{forms.F1099BWildcardBasis, forms.F1099BWildcardTerm},
				Compute: func(dv forms.DepValues) float64 {
					return dv.SumAllWhere(forms.F1099BWildcardBasis, forms.F1099BWildcardTerm, "long")
				},
			},
			// Total long-term wash sale adjustments
			{
				Line:      "lt_wash",
				Type:      forms.Computed,
				Label:     "Long-term wash sale adjustments",
				DependsOn: []string{forms.F1099BWildcardWashSale, forms.F1099BWildcardTerm},
				Compute: func(dv forms.DepValues) float64 {
					return dv.SumAllWhere(forms.F1099BWildcardWashSale, forms.F1099BWildcardTerm, "long")
				},
			},
			// Long-term gain or loss
			{
				Line:      "lt_gain_loss",
				Type:      forms.Computed,
				Label:     "Long-term gain or (loss)",
				DependsOn: []string{forms.F8949LTProceedsKey, forms.F8949LTBasisKey, forms.F8949LTWashKey},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.F8949LTProceedsKey) - dv.Get(forms.F8949LTBasisKey) + dv.Get(forms.F8949LTWashKey)
				},
			},
		},
	}
}

File: api/src/forms/federal/form_8949.rs
use crate::domain::field::*;
use crate::domain::form::*;

pub fn form_8949() -> FormDef {
    FormDef {
        id: FORM_F8949.to_string(),
        name: "Form 8949 — Sales and Other Dispositions of Capital Assets".to_string(),
        jurisdiction: Jurisdiction::Federal,
        tax_years: vec![2024, 2025, 2026],
        question_group: "income_1099".to_string(),
        question_order: 3,
        fields: vec![
            // --- Part I: Short-Term ---
            // Total short-term proceeds
            {
                let deps = vec![
                    F1099_B_WILDCARD_PROCEEDS.to_string(),
                    F1099_B_WILDCARD_TERM.to_string(),
                ];
                FieldDef::new_computed("st_proceeds", "Short-term total proceeds", deps, Box::new(|dv: &DepValues| {
                    dv.sum_all_where(F1099_B_WILDCARD_PROCEEDS, F1099_B_WILDCARD_TERM, "short")
                }))
            },
            // Total short-term cost basis
            {
                let deps = vec![
                    F1099_B_WILDCARD_BASIS.to_string(),
                    F1099_B_WILDCARD_TERM.to_string(),
                ];
                FieldDef::new_computed("st_basis", "Short-term total cost basis", deps, Box::new(|dv: &DepValues| {
                    dv.sum_all_where(F1099_B_WILDCARD_BASIS, F1099_B_WILDCARD_TERM, "short")
                }))
            },
            // Total short-term wash sale adjustments
            {
                let deps = vec![
                    F1099_B_WILDCARD_WASH_SALE.to_string(),
                    F1099_B_WILDCARD_TERM.to_string(),
                ];
                FieldDef::new_computed("st_wash", "Short-term wash sale adjustments", deps, Box::new(|dv: &DepValues| {
                    dv.sum_all_where(F1099_B_WILDCARD_WASH_SALE, F1099_B_WILDCARD_TERM, "short")
                }))
            },
            // Short-term gain or loss
            {
                let deps = vec![
                    F8949_ST_PROCEEDS.to_string(),
                    F8949_ST_BASIS.to_string(),
                    F8949_ST_WASH.to_string(),
                ];
                FieldDef::new_computed("st_gain_loss", "Short-term gain or (loss)", deps, Box::new(|dv: &DepValues| {
                    dv.get(F8949_ST_PROCEEDS) - dv.get(F8949_ST_BASIS) + dv.get(F8949_ST_WASH)
                }))
            },
            // --- Part II: Long-Term ---
            // Total long-term proceeds
            {
                let deps = vec![
                    F1099_B_WILDCARD_PROCEEDS.to_string(),
                    F1099_B_WILDCARD_TERM.to_string(),
                ];
                FieldDef::new_computed("lt_proceeds", "Long-term total proceeds", deps, Box::new(|dv: &DepValues| {
                    dv.sum_all_where(F1099_B_WILDCARD_PROCEEDS, F1099_B_WILDCARD_TERM, "long")
                }))
            },
            // Total long-term cost basis
            {
                let deps = vec![
                    F1099_B_WILDCARD_BASIS.to_string(),
                    F1099_B_WILDCARD_TERM.to_string(),
                ];
                FieldDef::new_computed("lt_basis", "Long-term total cost basis", deps, Box::new(|dv: &DepValues| {
                    dv.sum_all_where(F1099_B_WILDCARD_BASIS, F1099_B_WILDCARD_TERM, "long")
                }))
            },
            // Total long-term wash sale adjustments
            {
                let deps = vec![
                    F1099_B_WILDCARD_WASH_SALE.to_string(),
                    F1099_B_WILDCARD_TERM.to_string(),
                ];
                FieldDef::new_computed("lt_wash", "Long-term wash sale adjustments", deps, Box::new(|dv: &DepValues| {
                    dv.sum_all_where(F1099_B_WILDCARD_WASH_SALE, F1099_B_WILDCARD_TERM, "long")
                }))
            },
            // Long-term gain or loss
            {
                let deps = vec![
                    F8949_LT_PROCEEDS.to_string(),
                    F8949_LT_BASIS.to_string(),
                    F8949_LT_WASH.to_string(),
                ];
                FieldDef::new_computed("lt_gain_loss", "Long-term gain or (loss)", deps, Box::new(|dv: &DepValues| {
                    dv.get(F8949_LT_PROCEEDS) - dv.get(F8949_LT_BASIS) + dv.get(F8949_LT_WASH)
                }))
            },
        ],
    }
}
